File: src/two_sum_1/solution_binary_search.py
from typing import List


class Solution:
    def two_sum(self, nums: List[int], target: int) -> List[int]:
        nums_with_indices = sorted((num, idx) for idx, num in enumerate(nums))

        for idx, (num, original_idx) in enumerate(nums_with_indices):
            l = idx + 1
            r = len(nums_with_indices)

            while l < r:
                m = (l + r) // 2
                find = target - num

                if nums_with_indices[m][0] == find:
                    return [original_idx, nums_with_indices[m][1]]
                elif nums_with_indices[m][0] < find:
                    l = m + 1
                else:
                    r = m

        return [-1, -1]


def test():
    print(Solution().two_sum([3, 2, 4], 6))
    print(Solution().two_sum([0, 4, 3, 0], 0))

File: src/two_sum_1/solution_binary_search.rs
pub struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut nums_with_indices: Vec<(i32, usize)> = nums
            .iter()
            .enumerate()
            .map(|(idx, &num)| (num, idx))
            .collect();

        nums_with_indices.sort_by_key(|&(num, _)| num);

        for (idx, &(num, original_idx)) in nums_with_indices.iter().enumerate() {
            let mut l = idx + 1;
            let mut r = nums_with_indices.len();

            let find = target - num;
            while l < r {
                let m = (l + r) / 2;
                match nums_with_indices[m].0.cmp(&find) {
                    std::cmp::Ordering::Equal => {
                        return vec![original_idx as i32, nums_with_indices[m].1 as i32];
                    }
                    std::cmp::Ordering::Less => {
                        l = m + 1;
                    }
                    std::cmp::Ordering::Greater => {
                        r = m;
                    }
                }
            }
        }

        vec![-1, -1]
    }
}

pub fn test() {
    println!("{:?}", Solution::two_sum(vec![3, 2, 4], 6));
}
